rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,516 @@
+   but writes new data to a different set of SSTs from the original writer.
+
+## Goals
+
+In this RFC, we propose adding checkpoints and clones to address this limitation.
+- Clients will be able to establish cheap db checkpoints that reference the current version of SlateDB’s manifest.
+- Writers will use checkpoints to “pin” a version of the database for reads to avoid the GC from deleting the SSTs.
+- Readers will use checkpoints similarly, and many Reader processes can coexist alongside a Writer.
+- Finally, we will support creating so-called “clones” from a checkpoint to allow spawning new Writers that write to
+  a “fork” of the original database.
+
+### Out Of Scope
+
+This RFC does not propose:
+- Any mechanism to provide isolation guarantees from other concurrent reads/writes to the database.
+
+## Proposal
+
+At a high level, this proposal covers a mechanism for establishing database checkpoints, and reviews how the various
+processes will interact with this mechanism. Logically, a checkpoint refers to a consistent and durable view of the
+database at some point in time. It is consistent in the sense that data read from a checkpoint reflects all writes up
+to some point in time, and durable in the sense that this property holds across process restarts. Physically, a
+checkpoint refers to a version of SlateDB’s manifest.
+
+### Manifest Schema
+
+Checkpoints themselves are also stored in the manifest. We’ll store them using the following schema:
+
+```
+
+table DbParent {
+   // Optional path to a parent database from which this database was cloned
+   parent_path: string (required);
+
+   // Optional parent checkpoint ID
+   parent_checkpoint: UUID (required);
+}
+
+table ManifestV1 {
+   // Optional details about the parent checkpoint for the database
+   parent: DbParent
+
+   // Flag to indicate whether initialization has finished. When creating the initial manifest for
+   // a root db (one that is not a clone), this flag will be set to true. When creating the initial
+   // manifest for a clone db, this flag will be set to false and then updated to true once clone
+   // initialization has completed.
+   initialized: boolean
+
+   // Optional epoch time in seconds that this database was destroyed
+   destroyed_at_s: u64;
+
+   ...
+
+   // A list of current checkpoints that may be active (note: this replaces the existing
+   // snapshots field)
+   checkpoint: [Checkpoints] (required);
+}
+
+
+table WriterCheckpoint {
+   epoch: u64;
+}
+
+union CheckpointMetadata { WriterCheckpoint }
+
+table Uuid {
+   high: uint64;
+   low: uint64;
+}
+
+
+// Checkpoint reference to be included in manifest to record active checkpoints.
+table Checkpoint {
+   // Id that must be unique across all open checkpoints.
+   id: Uuid (required);
+
+   // The manifest ID that this checkpoint is using as its `CoreDbState`.
+   manifest_id: ulong;
```

Comment:
During the manifest RFC, I think we had a comment about `manifest_id` being the id of the current manifest, and only current. i.e. We cannot create a checkpoint for an older manifest.

Is that still the restriction?